network/p2p/gossip/metrics.go

```diff
@@ -0,0 +1,185 @@
+	}
+	typeLabels   = []string{typeLabel}
+	unsentLabels = prometheus.Labels{
+		typeLabel: unsentType,
+	}
+	sentLabels = prometheus.Labels{
+		typeLabel: sentType,
+	}
+)
+
+// Metrics that are tracked across a gossip protocol. A given protocol should
+// only use a single instance of Metrics.
+type Metrics struct {
+	count                   *prometheus.CounterVec
+	bytes                   *prometheus.CounterVec
+	tracking                *prometheus.GaugeVec
+	trackingLifetimeAverage prometheus.Gauge
+	topValidators           *prometheus.GaugeVec
+
+	labelsLock        sync.Mutex
+	gossipablesLabels map[ids.ID]string
+}
+
+// NewMetrics returns a common set of metrics
+func NewMetrics(
+	metrics prometheus.Registerer,
+	namespace string,
+) (*Metrics, error) {
+	m := &Metrics{
+		count: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Name:      "gossip_count",
+				Help:      "amount of gossip (n)",
+			},
+			ioTypeDroppedLabels,
+		),
+		bytes: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Name:      "gossip_bytes",
+				Help:      "amount of gossip (bytes)",
+			},
+			ioTypeDroppedLabels,
+		),
+		tracking: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: namespace,
+				Name:      "gossip_tracking",
+				Help:      "number of gossipables being tracked",
+			},
+			typeLabels,
+		),
+		trackingLifetimeAverage: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "gossip_tracking_lifetime_average",
+			Help:      "average duration a gossipable has been tracked (ns)",
+		}),
+		topValidators: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: namespace,
+				Name:      "top_validators",
+				Help:      "number of validators gossipables are sent to due to stake",
+			},
+			typeLabels,
+		),
+		gossipablesLabels: make(map[ids.ID]string),
+	}
+	err := errors.Join(
+		metrics.Register(m.count),
+		metrics.Register(m.bytes),
+		metrics.Register(m.tracking),
+		metrics.Register(m.trackingLifetimeAverage),
+		metrics.Register(m.topValidators),
+	)
+	return m, err
+}
+
+func (m *Metrics) observeMessage(labels prometheus.Labels, count int, bytes int) error {
+	countMetric, err := m.count.GetMetricWith(labels)
+	if err != nil {
+		return fmt.Errorf("failed to get count metric: %w", err)
+	}
+
+	bytesMetric, err := m.bytes.GetMetricWith(labels)
+	if err != nil {
+		return fmt.Errorf("failed to get bytes metric: %w", err)
+	}
+
+	countMetric.Add(float64(count))
+	bytesMetric.Add(float64(bytes))
+	return nil
+}
+
+// ObserveIncomingGossipable allows external-package mempools to label a custom gossipable drop-reason and have it
+// included in the gossip metrics.
+func (m *Metrics) ObserveIncomingGossipable(txID ids.ID, label string) {
+	m.labelsLock.Lock()
+	defer m.labelsLock.Unlock()
+	_, ok := m.gossipablesLabels[txID]
+	if ok {
+		// we already have this gossipable labeled; don't re-label.
```

Comment:
This seems like a confusing ux to me... why handle de-duping for the caller? If the caller wants de-duped behavior, they can just check the error returned from `Add` before adding their metric. At the very least if we kept this behavior it should be explicitly documented.

Reply:
this behavior was explicitly meant for the "Others" group. Given that we don't have it anymore, I'll remove it.